mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala

```diff
@@ -89,10 +89,14 @@ class MatrixFactorizationModel(
    */
   def predict(usersProducts: RDD[(Int, Int)]): RDD[Rating] = {
     val users = userFeatures.join(usersProducts).map {
-      case (user, (uFeatures, product)) => (product, (user, uFeatures))
+      case (user, (uFeatures, product)) => ((user, product), uFeatures)
     }
-    users.join(productFeatures).map {
-      case (product, ((user, uFeatures), pFeatures)) =>
+    val productUsers = usersProducts.map(up => (up._2, up._1))
+    val products = productFeatures.join(productUsers).map {
```

Comment:
Yes. That is what I thought at the first. Also it will add the additional cost of calculating distinct numbers of users and products. Since the input parameter is a RDD of (userId, productId), to calculate that may require two distinct + count? Is it costing less than another join?